### ISSUE TITLE
CA1508 False positive reporting

### ISSUE THREAD

**Issue by RamjotSingh** [dotnet/roslyn-analyzers#3874]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.3.0-beta2.final

#### Diagnostic ID

Example: [CA1508](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1508)

#### Repro steps

Attaching the file below which has 3 repros with issues
[CA1508IssuesRepro.zip](https://github.com/dotnet/roslyn-analyzers/files/4933276/CA1508IssuesRepro.zip)

#### Expected behavior
CA1508 should not fire

#### Actual behavior
CA1508 fires

##### Code snippets

```cs
            for (int i = 0; i < 100; i++)
            {
                try
                {
                    File.ReadAllText("C:\\pagefile.sys");
                    break;
                }
                catch (Exception ex)
                {
                    if (i == 98)
                    {
                        // CodeAnalysis thinks this lines cannot be reached.
                        throw;
                    }

                    // Code Analysis thinks following condition cannot be true. However if we change it to ex is IOException it will work.
                    if (ex as IOException != null)
                    {
                        Thread.Sleep(TimeSpan.FromSeconds(1));
                    }
                }
            }
```

```cs
        static void DerivedVsBaseClass(JToken data)
        {
            if (data == null)
            {
                return;
            }

            // CA1508 thinks following condition can't be true. When we know it can be!
            if (data as JArray != null)
            {
                throw new ArgumentException("JArray not allowed");
            }
        }
```

```cs
        static void HeaderDictionaryReads()
        {
            IHeaderDictionary headerDictionary = new HeaderDictionary();

            headerDictionary["CorrectHeader"] = "Value";

            string wrongHeader = headerDictionary["IncorrectHeader"];

            // CA1508 thinks the following cannot be true. While it obviously will be in the case below. StringValues.IsNullOrEmpty would work though.
            if(string.IsNullOrEmpty(wrongHeader))
            {
                throw new ArgumentException("AsIntended");
            }
        }
```

**Comment by SebastianSchumann** [comment]
After Update to 3.3.0 (from 3.0.0) the problem is not reported for the positive case:

    public static void Foo<TKey, TValue>(this IDictionary<TKey, IEnumerable<TValue>> self)
    {
        foreach (var (key, value) in self)
        {
            if (value is null)
            {
                continue;
            }

            // CA1508 reports that this condition is always true.
            if (!(value is ReadOnlyCollection<TValue> readOnlyValue))
            {
            }
        }
    }

If the null check is removed or the condition is inverted to `if (value is ReadOnlyCollection<TValue> readOnlyValue)` CA1508 is not reported.

**Comment by mavasani** [comment]
First false positive still repros on latest bits in main branch.
